kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt

```diff
@@ -0,0 +1,760 @@
+      if (rawType == ARRAY) {
+        return CodeBlock.of("%T.arrayOf(%L)",
+            Types::class,
+            typeArguments[0].objectType().makeType(elementUtils, typesArray, genericTypeNames))
+      }
+      // If it's a Class type, we have to specify the generics.
+      val rawTypeParameters = if (rawType.isClass(elementUtils)) {
+        CodeBlock.of(
+            typeArguments.joinTo(
+                buffer = StringBuilder(),
+                separator = ", ",
+                prefix = "<",
+                postfix = ">") { "%T" }
+                .toString(),
+            *(typeArguments.map { objectType() }.toTypedArray())
+        )
+      } else {
+        CodeBlock.of("")
+      }
+      CodeBlock.of(
+          "%T.newParameterizedType(%T%L::class.java, ${typeArguments
+              .joinToString(", ") { "%L" }})",
+          Types::class,
+          rawType.objectType(),
+          rawTypeParameters,
+          *(typeArguments.map {
+            it.objectType().makeType(elementUtils, typesArray, genericTypeNames)
+          }.toTypedArray()))
+    }
+    is WildcardTypeName -> {
+      val target: TypeName
+      val method: String
+      when {
+        lowerBounds.size == 1 -> {
+          target = lowerBounds[0]
+          method = "supertypeOf"
+        }
+        upperBounds.size == 1 -> {
+          target = upperBounds[0]
+          method = "subtypeOf"
+        }
+        else -> throw IllegalArgumentException(
+            "Unrepresentable wildcard type. Cannot have more than one bound: " + this)
+      }
+      CodeBlock.of("%T.%L(%T::class.java)", Types::class, method, target)
+    }
+    is TypeVariableName -> {
+      CodeBlock.of("%N[%L]", typesArray, genericTypeNames.indexOfFirst { it == this })
+    }
+    else -> throw IllegalArgumentException("Unrepresentable type: " + this)
+  }
+}
+
+private data class Property(
+    val name: String,
+    val fqClassName: String,
+    val serializedName: String,
+    val hasDefault: Boolean,
+    val nullable: Boolean,
+    val typeName: TypeName,
+    val jsonQualifiers: Set<AnnotationMirror>) {
+
+  val isRequired = !nullable && !hasDefault
+}
+
+private data class Adapter(
+    val fqClassName: String,
+    val packageName: String,
+    val propertyList: List<Property>,
+    val originalElement: Element,
+    val name: String = fqClassName.substringAfter(packageName)
+        .replace('.', '_')
+        .removePrefix("_"),
+    val hasCompanionObject: Boolean,
+    val visibility: Visibility,
+    val elementUtils: Elements,
+    val genericTypeNames: List<TypeVariableName>?) {
+
+  fun generate(adapterName: String, fileSpecBuilder: FileSpec.Builder) {
+    val nameAllocator = NameAllocator()
+    fun String.allocate() = nameAllocator.newName(this)
+
+    val originalTypeName = originalElement.asType().asTypeName()
+    val moshiName = "moshi".allocate()
+    val moshiParam = ParameterSpec.builder(moshiName, Moshi::class.asClassName()).build()
```

Comment:
`asClassName()` looks redundant, you should be able to simply pass `Moshi::class`

Reply:
9a34bb0